### PR TITLE
[WIP] Use Python logging module for logging

### PR DESCRIPTION
**Pull request by tjikkun**
This pull request is to get some feedback.
The purpose of this branch is to switch to the Python built-in logging framework instead of the custom debugLog function. It is not fully ready yet, but file and CSV logging work.
A big upside is that with the logging framework it is easy to integrate with Sentry and get good traces.
Also, if TWCManager is used as a module, logging will work as expected.
Please let me know what you think, if it is something you would accept, or if I should stop wasting my time 😃

### CODE REVIEW

**Comment by ngardiner** [comment]
It is not something I would be looking to merge, sorry to say. We have put a lot of effort into modularising logging and making it context aware, so it can collect details and values from the various modules and store them in appropriate structures (hence the CSV and Database modules, which I actively use to record charging activity) and I suspect the assumption here is that we are just logging log messages straight to these formats but that is not the case, we in fact take the values reported and massage them into log messages or database tables as appropriate for the data types.

A generic log module would not allow us to do this. You can take a look at the MySQL/SQLite/CSV log modules and see what I mean, some logs are text only such as those raised by debugLog, but for logs with structured data we write only the values to the DB as writing entire log strings as a generic log module would do offers little value for recording and querying data.

The python logging module approach, while it might be seen as a standard for logging, still introduces an external dependency, but then immediately loses functionality such as the reporting of which module raised any log message which is pretty much essential in this architecture, so it doesn't seem like a valuable trade off to me.

Rather than taking this approach, what I would highly encourage is to improve the current logging architecture to meet the requirements you had when introducing this change, so we don't lose functionality or fidelity of the data we collect and we work with additional use cases. I'm quite confident we could support the use cases you mention using the inbuilt log modules, extend them to be even better and not lose functionality that we have today - we don't need to adopt a generic/standard log module if it gives us less than we have today.

Do you know what is missing from the current logging modules that would stop you integrating with Sentry or running as a module? Not only would I expect both to work as is, I would expect that it would be easier as is to integrate with external interfaces as we log structured data only for CSV/SQLite/etc and you would not need to do any log parsing in the current system.

**Comment by tjikkun** [comment]
Hi, thanks for taking the time to respond!
> I suspect the assumption here is that we are just logging log messages straight to these formats but that is not the case, we in fact take the values reported and massage them into log messages or database tables as appropriate for the data types.

This is very much not the case! I saw this was the case, and this was the reason for me to spend much time, before even asking for input: I wanted to make sure that no functionality or flexibility would be lost. the logging module provides support for structured data, using extra= on the log call. I have tested this on the CSV module and it work great, just as it is now.

The python logging module is already being used in FileLogging.py right now, so it is not an extra dependency.

For sentry support, adding that is just:
```
import sentry_sdk

sentry_sdk.init(
    "https://examplePublicKey@o0.ingest.sentry.io/0",

    # Set traces_sample_rate to 1.0 to capture 100%
    # of transactions for performance monitoring.
    # We recommend adjusting this value in production.
    traces_sample_rate=1.0,
)
```
But, this assumes you are using the logging module.

Sentry gets context from the place where the logging call is being made, so that's why I thought it made sense to create this PR.
By the way, I am in no way suggesting we should remove the logging modules, instead, I propose to change from:

in a module, call master.debugLog which in turn loops through all modules, calling module.debugLog
to:
in a module, call logger.log. All logging modules receive the log.

and, for structured data (green energy as example) change from:
in TWCManager.py, loop through all modules, calling greenEnergy(data)
to:
in TWCManager.py call logger.log with extra=data (log_type: "green_energy"). All logging modules receive this, and can decide to use this structured data or not.


So, to summarize:
I don't want to throw out the great logging modules that are there, just tweak the way they are being called. This gives greater compatibility with sentry for example, but also makes it easy to catch logging from any modules we use, like serial.

I hope you will reconsider, if you do, I will, I will finish this PR, taking great care not to change anything. Of course, if you then don't like the end result you can still reject, but I won't continue on this if there is not at least a chance of approval.

**Comment by ngardiner** [comment]
I see your point @tjikkun, and I appreciate you've done more analysis of this proposal than I had given you credit for.

Let me run up your changes and observe, you've put a significant amount of effort in and I'd like to be better informed on how this change would look and feel.

**Comment by tjikkun** [comment]
You can wait a bit, I'll put it in a little better state, so you can judge it better.
I think I'll get to it this weekend to have at least the CVS module "fully" working, and clean it up a little so you can really see where I would like to take it, there are some ugly sots at the moment
Thanks for being open to this, I appreciate the thought and I must say I like how you maintain this project. Even if, in the end, you decide you don't like the change in the PR :)

**Comment by tjikkun** [comment]
Updated, this should give a good idea of the direction I am going. MySQLLogging.py is not currently updated, but CVSLogging should work (maybe with some duplicates because there is still more work to do). I haven't been able to test with a car attached yet, but at least green energy is tested and works fine.

**Comment by tjikkun** [comment]
Now SQLite should work, and also MySQL (but MySQL is not tested yet by me).
So all in all this should be in okayish state now, at least for a review. Hope you will find some time to look at it, and give me your opinion.
Of course I will clean up commits, if you decide to take this change.

**Comment by ngardiner** [comment]
Thanks @tjikkun, I'm running this up in my environment now, and I am using MySQL logging - will get back with whether MySQL logging is working.

**Comment by ngardiner** [comment]
Hi @tjikkun, thanks for your patience, I am going to merge this now. I'm confident everything is working under the python logging module and you've taken care of anything I could have expected such as the mapping from debugLevel to logLevel in config, It's a solid piece of work that would have taken a significant amount of work and which simplifies the use of logging modules for us. Thank you!

**Comment by tjikkun** [comment]
Cool, thanks for merging!
Have you tried out the sentry integration yet? If not, you should really give it a go :)

**Comment by ngardiner** [comment]
Running it up in my dev environment now, it's already very useful as it has picked up a mysql query that fails intermittently, so it's great. One question I have for you @tjikkun, do you think the performance monitoring feature of Sentry would be useful for tracking any of the values we report? I am not all that familiar with what it tracks/shows, so not sure if it would be relevant or not

**Comment by tjikkun** [comment]
@ngardiner I haven't really thought about it before. It usually is used for measuring the speed of generating web pages. I don't think it maps to our reporting. That said, I am not an expert on this, so someone might prove me wrong :)
In any case, I am happy that it already has proven useful. Especially for intermittent errors, sentry is a great tool.
